### ISSUE TITLE
Feature Request: Send cloud push notifications to multiple devices

### ISSUE THREAD

**Issue by anik786** [kreait/firebase-php#261]
Can we allow push notification to a batch of device IDs, instead of one at a time.

This seems that it is possible:

https://firebase.google.com/docs/cloud-messaging/android/device-group
and
https://firebase.google.com/docs/cloud-messaging/http-server-ref#send-downstream

**Comment by jeromegamez** [comment]
Yes, unfortunately it's not (yet) implemented in the HTTP v1 API, and the message formats are not compatible... I've written a little about it in #192.

If you want, I can keep this issue open as a future reference, but I still can't give an ETA on that feature, or if it can be implemented without non-legacy API support at all.

**Comment by jeromegamez** [comment]
However, I believe that the recommended way of sending messages is to do it via topics instead of multiple device ids at once (I'm not willing to argue, though, because I'm really not sure 😅

**Comment by anik786** [comment]
I thought as much. Would it be possible to build a solution that uses the legacy API part for this very important feature.

Topics has soo many issues with it. For example, it's hard to keep it in sync with the database. It's also hard to do more complex logic without creating many dummy topics. You also cannot delete the topic, without individually unsubscribing each user. This becomes even more problematic if you miss a FCM token for whatever reason. There also seems to be very little resources explaining how to do basic stuff with topics, like listing all the tokens for a topic. I could go on, but sending notifications to devices seems the best way to do this.

**Comment by anik786** [comment]
Looks like mix and matching between these different API's is not that difficult. For the above API call the only additional information you need is the server key, and then the only main difference is the API URL.

**Comment by jeromegamez** [comment]
I just released [4.18.0](https://github.com/kreait/firebase-php/releases/tag/4.18.0) which allows you to re-use a composed message with another target: https://firebase-php.readthedocs.io/en/latest/cloud-messaging.html#changing-the-message-target

It's not optimal but is the closest we can currently get without breaking backwards compatibility in the 4.x branch. It would work like this:

```php
$deviceIds = []; // array with deviceids
$message = CloudMessage::withTarget('token', 'foo') // The device ID doesn't matter here
    ->withNotification(['title' => 'My Title', 'body' => 'My body']);

foreach ($deviceIds as $deviceId) {
    $messaging->send($message->withChangedTarget('token', $deviceId));
}
```

**Comment by kublaios** [comment]
Thanks for the workaround @jeromegamez. Still would be great if this gets implemented though.

**Comment by jeromegamez** [comment]
I still think that what I wrote in https://github.com/kreait/firebase-php/issues/261#issuecomment-433530152 would be the optimal way to solve the problem (this time I'll add a code example 😅)

```php
$messaging = $firebase->getMessaging();

$topic = uniqid('topic_', false);
$deviceIds = [/* ... */]; // array with deviceids

$messaging->subscribeToTopic($topic, $deviceIds);

$message = CloudMessage::withTarget('topic', $topic)
    ->withNotification(['title' => 'Notification title', 'body' => 'Notification body']);

$messaging->send($message);

$messaging->unsubscribeFromTopic($topic, $deviceIds);
```

I don't think that this is less comfortable then sending to a device group - in fact it's even documented like this in e.g. the official Firebase for Android documentation at https://firebase.google.com/docs/cloud-messaging/android/send-multiple#subscribe_the_client_app_to_a_topic 

You get all the [benefits of using the newer API](https://firebase.google.com/docs/cloud-messaging/migrate-v1) without the [limitations of device groups](https://firebase.google.com/docs/cloud-messaging/ios/device-group).

**Comment by jeromegamez** [comment]
After having researched this again, I grow more and more confident that this is the way to go, and adding a bridge to the legacy messaging API would add... legacy to the SDK that I wouldn't want to maintain :). So, I will close this issue now, but feel free to add further comments if you wish!

**Comment by jeromegamez** [comment]
Oh, Look what Google released just a couple of hours ago: https://firebase.google.com/support/release-notes/admin/node#7.1.0

I’m re-opening this feature request, I probably can extract the functionality :)

**Comment by sakhdevel** [comment]
Hi! Thanks for this project :+1: 
I've got two questions.

1.
> I still think that what I wrote in #261 (comment) would be the optimal way to solve the problem (this time I'll add a code example)

It is an interesting solution, but I worry about the frequency of topics adding/removing (since, topics probably were not supposed to be used this way). Are you sure that nothing will be blocked? Let's say a project creates/removes topics several times per minute. How do you think, can one use this solution in production?

2. 
> https://firebase.google.com/support/release-notes/admin/node#7.1.0
> I’m re-opening this feature request, I probably can extract the functionality :)

I didn't manage to find an example on how to do this via REST, only examples for Java and other languages. Is it possible to try this out in Postman?

P.S.: It would be great to see this feature in your project :+1:

**Comment by sakhdevel** [comment]
Original question was 

> Can we allow push notification to a batch of device IDs

Should I create another feature request for pushing **several notifications** to **several tokens** per one request? This is called "messaging.sendAll()" in the link you provided

**Comment by Arkemlar** [comment]
Following [this](https://firebase.google.com/docs/cloud-messaging/send-message#send_messages_to_multiple_devices) chapter of FCM documentation it is quite easy to send message to multiple devices. In context of this library it just requires some changes in MessageTarget class...

**Comment by jeromegamez** [comment]
I'm working on it! In the meantime, release [4.22](https://github.com/kreait/firebase-php/releases) opens the `CloudMessage` a little, so that you can create a message without having to specify a target at creation time. It's not documented in the docs yet, but here is how you could implement a poor man's multicast in your application:

```php
$tokens = ['token-a', 'token-b', 'token-c']; // Not actual tokens
$message = CloudMessage::new();

foreach ($tokens as $token) {
    $messaging->send($message->withChangedTarget('token', $token);
}
```

**Comment by jeromegamez** [comment]
I just released [4.24.0](https://github.com/kreait/firebase-php/releases) which adds `Messaging::sendMulticast($message, $deviceTokens): MulticastSendReport`.

This saves you from looping over the device tokens by yourselves, but still has the drawback mentioned in https://github.com/kreait/firebase-php/issues/261#issuecomment-434739055 that this will result in one request per recipient.

This is not optimal, but I didn't want to keep you waiting, and now that the method exists, I can ensure that the tests pass when I refactor the internals to use the Firebase batch API.

I created #302 as a follow up issue for this.
